c2rust-analyze/src/dataflow/type_check.rs

```diff
@@ -142,7 +148,59 @@ impl<'tcx> TypeChecker<'tcx, '_> {
             Rvalue::Len(pl) => {
                 self.visit_place(pl, Mutability::Not);
             }
-            Rvalue::Cast(_, ref op, _) => self.visit_operand(op),
+            Rvalue::Cast(CastKind::PointerFromExposedAddress, ref op, _) => {
+                // We support only one case here, which is the case of null pointers
+                // constructed via casts such as `0 as *const T`
+                if let Some(true) = op.constant().cloned().map(util::is_null_const) {
+                    self.visit_operand(op)
+                } else {
+                    panic!("Creating non-null pointers from exposed addresses not supported");
+                }
+            }
+            Rvalue::Cast(CastKind::Pointer(PointerCast::Unsize), ref op, ty) => {
+                let pointee_ty = match *ty.kind() {
+                    TyKind::Ref(_, ty, _) => ty,
+                    TyKind::RawPtr(tm) => tm.ty,
+                    _ => unreachable!("unsize cast has non-pointer output {:?}?", ty),
+                };
+
+                let op_lty = self.acx.type_of(op);
+                assert!(matches!(
+                    op_lty.kind(),
+                    TyKind::Ref(..) | TyKind::RawPtr(..)
+                ));
+                assert_eq!(op_lty.args.len(), 1);
+                let op_pointee_lty = op_lty.args[0];
+
+                if let TyKind::Slice(elem_ty) = *pointee_ty.kind() {
+                    assert!(matches!(op_pointee_lty.kind(), TyKind::Array(..)));
+                    assert_eq!(op_pointee_lty.args.len(), 1);
+                    let elem_lty = op_pointee_lty.args[0];
+                    assert_eq!(elem_lty.ty, elem_ty);
+                    assert_eq!(op_pointee_lty.label, PointerId::NONE);
+                    self.do_assign_pointer_ids(rvalue_lty.label, op_lty.label);
+                }
+
+                self.visit_operand(op)
+            }
+            Rvalue::Cast(CastKind::Pointer(..), ref op, _) => {
+                let op_lty = self.acx.type_of(op);
+
+                // The source and target types are both pointers, and they have identical
+                // pointee types.
+                // TODO: remove or move check to `is_castable_to`
+                assert!(op_lty.args[0].ty == rvalue_lty.args[0].ty);
+
+                assert!(is_castable_to(op_lty, rvalue_lty));
+                self.visit_operand(op)
+            }
+            Rvalue::Cast(_cast_kind, ref op, _) => {
+                // A cast such as `T as U`
+                let casted_from = self.acx.type_of(op);
+                let casted_to = rvalue_lty;
+                assert!(is_castable_to(casted_from, casted_to));
+                self.visit_operand(op)
+            }
```

Comment:
Ptr and non-ptr casts are quite different; I don't think there's really any `is_castable_to` function we could conceivably put here.  Non-ptr casts are going to be things like scalar casts without any pointers, so we don't need to really care about those, right, as there are no pointers?  Even if it's some sort of transmuting cast not going through a pointer, `rustc` typeck isn't going to allow that in the first place.  Ptr-to-ptr casts are the main thing we care about here, and that's how I designed the #839 and #841 PRs.

Reply:
I don't think there is harm in keeping the check here and making it trivially true for non-ptr types. Do you?

Reply:
`CastKind::PointerExposeAddress` will fall into this case at the moment, so it's at least possible for something nontrivial to happen here.

Reply:
Marking resolved. Feel free to mark unresolved with specific changes to request

Reply:
I'll figure it out when I get to #839 and #841.